File: Spend Management/shared/code/cHelp.cs
using System;
using System.Configuration;
using System.Linq;
using System.Web.Caching;
using System.Data;
using System.Collections.Generic;
using SpendManagementLibrary;


namespace Spend_Management
{
	/// <summary>
	/// Summary description for cHelp.
	/// </summary>
	public class cHelp
	{

		public SortedList<Guid, cHelpItem> list;
		
		public System.Web.Caching.Cache Cache = (System.Web.Caching.Cache)System.Web.HttpContext.Current.Cache;
        private int nAccountid;

		string strsql;

        public cHelp()
        {
            InitialiseGeneralData();
        }

        private void InitialiseGeneralData()
        {
            list = (SortedList<Guid, cHelpItem>)Cache["help"];
            if (list == null)
            {
                list = CacheGeneralList();
            }
        }

        private SortedList<Guid, cHelpItem> CacheGeneralList()
        {
            DBConnection expdata = new DBConnection(ConfigurationManager.ConnectionStrings["metabase"].ConnectionString);
            SortedList<Guid, cHelpItem> list = new SortedList<Guid, cHelpItem>();            
            Guid tooltipID;
            string helptext, description, page, tooltipArea;
            int? moduleID;
            cHelpItem reqhelp;
            System.Data.SqlClient.SqlDataReader reader;

            strsql = "select page, description, helptext, tooltipID, tooltipArea, moduleID from dbo.help_text";
            expdata.sqlexecute.CommandText = strsql;
            if (GlobalVariables.GetAppSettingAsBoolean("EnableBrokers"))
            {
                var dep = new SqlCacheDependency(expdata.sqlexecute);
                Cache.Insert("help", list, dep, System.Web.Caching.Cache.NoAbsoluteExpiration, TimeSpan.FromMinutes((int)Caching.CacheTimeSpans.Permanent), CacheItemPriority.Default, null);
            }

            using (reader = expdata.GetReader(strsql))
            {
                while (reader.Read())
                {                    
                    if (reader.IsDBNull(reader.GetOrdinal("page")) == true)
                    {
                        page = "";
                    }
                    else
                    {
                        page = reader.GetString(reader.GetOrdinal("page"));
                    }
                    if (reader.IsDBNull(reader.GetOrdinal("helptext")) == true)
                    {
                        helptext = "";
                    }
                    else
                    {
                        helptext = reader.GetString(reader.GetOrdinal("helptext"));
                        //helptext = helptext.Replace("'", "\\'").Replace("\\\\'", "\\'");
                        //helptext = helptext.Replace("\"", "&quot;");
                    }

                    tooltipArea = reader.GetString(reader.GetOrdinal("tooltipArea"));
                    tooltipID = reader.GetGuid(reader.GetOrdinal("tooltipID"));

                    if (reader.IsDBNull(reader.GetOrdinal("moduleID")) == false)
                    {
                        moduleID = reader.GetInt32(reader.GetOrdinal("moduleID"));
                    }
                    else
                    {
                        moduleID = null;
                    }

                    description = reader.GetString(reader.GetOrdinal("description"));                    

                    reqhelp = new cHelpItem(page, description, helptext, tooltipID, tooltipArea, moduleID);

                    list.Add(tooltipID, reqhelp);
                }

                reader.Close();
            }

            return list;
        }
		public cHelp(int accountid)
		{
            nAccountid = accountid;
			InitialiseData();
        }

        #region properties
        public int accountid
        {
            get { return nAccountid; }
        }
        #endregion
        private void InitialiseData()
		{
            list = (SortedList<Guid, cHelpItem>)Cache["help" + accountid];
			if (list == null)
			{
				list = CacheList();
			}
		}

        /// <summary>
        /// Force the cache to reset
        /// </summary>
        private void ResetCache()
        {
            Cache.Remove("help" + accountid);
            list = null;
            InitialiseData();
        }

        private SortedList<Guid, cHelpItem> CacheList()
		{
            DBConnection expdata = new DBConnection(cAccounts.getConnectionString(accountid));
            SortedList<Guid, cHelpItem> list = new SortedList<Guid, cHelpItem>();			
            Guid tooltipID;
			string helptext, description, page, tooltipArea;
            int? moduleID;
			cHelpItem reqhelp;
			System.Data.SqlClient.SqlDataReader reader;

            AggregateCacheDependency aggdep = null;
            if (GlobalVariables.GetAppSettingAsBoolean("EnableBrokers"))
            {
                DBConnection metadata =
                    new DBConnection(ConfigurationManager.ConnectionStrings["metabase"].ConnectionString);
                strsql = "select page, description, helptext, tooltipID, tooltipArea, moduleID from dbo.help_text";
                metadata.sqlexecute.CommandText = strsql;
                SqlCacheDependency metadep = metadata.CreateSQLCacheDependency(strsql, new SortedList<string, object>());

                DBConnection depdata = new DBConnection(cAccounts.getConnectionString(accountid));
                strsql =
                    "select page, description, helptext, tooltipID, tooltipArea, moduleID from dbo.customised_help_text";
                depdata.sqlexecute.CommandText = strsql;
                SqlCacheDependency dep = depdata.CreateSQLCacheDependency(strsql, new SortedList<string, object>());
                aggdep = new AggregateCacheDependency();
                aggdep.Add(new CacheDependency[] {metadep, dep});
            }
            strsql = "select page, description, helptext, tooltipID, tooltipArea, moduleID from dbo.help_text";

            using (reader = expdata.GetReader(strsql))
            {
                while (reader.Read())
                {
                    page = reader.IsDBNull(reader.GetOrdinal("page")) ? string.Empty : reader.GetString(reader.GetOrdinal("page"));

                    helptext = reader.IsDBNull(reader.GetOrdinal("helptext")) ? string.Empty : reader.GetString(reader.GetOrdinal("helptext"));

                    tooltipArea = reader.GetString(reader.GetOrdinal("tooltipArea"));
                    tooltipID = reader.GetGuid(reader.GetOrdinal("tooltipID"));

                    description = reader.GetString(reader.GetOrdinal("description"));

                    if (reader.IsDBNull(reader.GetOrdinal("moduleID")) == false)
                    {
                        moduleID = reader.GetInt32(reader.GetOrdinal("moduleID"));
                    }
                    else
                    {
                        moduleID = null;
                    }

                    reqhelp = new cHelpItem(page, description, helptext, tooltipID, tooltipArea, moduleID);

                    list.Add(tooltipID, reqhelp);
                }

                reader.Close();
            }

            if (aggdep != null)
            {
                Cache.Insert("help" + this.accountid, list, aggdep, Cache.NoAbsoluteExpiration,
                    TimeSpan.FromMinutes((int) Caching.CacheTimeSpans.Permanent), CacheItemPriority.Default, null);
            }

			return list;
		}

        public cHelpItem GetToolTipByIntAndArea(string contextKey)
        {
            string[] temp = contextKey.Split(',');

            Guid tooltipId = Guid.Parse(temp[0]);

            string tooltipArea = temp[1];

            return
                this.list.Values.FirstOrDefault(
                    help => help.ToolTipID == tooltipId && tooltipArea.ToLower() == help.ToolTipArea.ToLower());
        }

		public cHelpItem getHelpById(Guid tooltipID)
		{
            cHelpItem item = null;
            list.TryGetValue(tooltipID, out item);
            return item;
		}

        public DataTable getGrid(int activeModuleID)
        {
            var tbl = new DataTable();            
            tbl.Columns.Add("page", typeof(string));
            tbl.Columns.Add("description", typeof(string));
            tbl.Columns.Add("text", typeof(string));
            tbl.Columns.Add("tooltipID", typeof(Guid));
            tbl.Columns.Add("tooltipArea", typeof(string));

            foreach (cHelpItem item in this.list.Values)
            {
                if (item.ModuleID.HasValue == false || (item.ModuleID.HasValue && activeModuleID == item.ModuleID.Value))
                {
                    if (item.page.ToLower() != "logon" && item.page.ToLower() != "forgotten password" && item.page.ToLower() != "self registration")
                    {
                        tbl.Rows.Add(
                            new object[]
                                {                                    
                                    item.page, item.description, item.helptext, item.ToolTipID,
                                    item.ToolTipArea
                                });
                    }
                }
            }
            return tbl;
        }

        public bool containsTip(Guid id)
        {
            cHelpItem item;


            list.TryGetValue(id, out item);
            if (item == null)
            {
                return false;
            }

            if (item.helptext.Trim().Length == 0)
            {
                return false;
            }

            return true;
        }

        public void saveTooltip(Guid tooltipID, string text)
        {
            DBConnection expdata = new DBConnection(cAccounts.getConnectionString(accountid));

            expdata.sqlexecute.Parameters.AddWithValue("@text", text);
            expdata.sqlexecute.Parameters.AddWithValue("@tooltipID", tooltipID);
            CurrentUser currentUser = cMisc.GetCurrentUser();
            expdata.sqlexecute.Parameters.AddWithValue("@CUemployeeID", currentUser.EmployeeID);
            if (currentUser.isDelegate == true)
            {
                expdata.sqlexecute.Parameters.AddWithValue("@CUdelegateID", currentUser.Delegate.EmployeeID);
            }
            else
            {
                expdata.sqlexecute.Parameters.AddWithValue("@CUdelegateID", DBNull.Value);
            }
            expdata.ExecuteProc("saveTooltip");
            expdata.sqlexecute.Parameters.Clear();
        }

        public void restoreTooltip(Guid tooltipID)
        {
            DBConnection expdata = new DBConnection(cAccounts.getConnectionString(accountid));

            expdata.sqlexecute.Parameters.AddWithValue("@tooltipID", tooltipID);
            CurrentUser currentUser = cMisc.GetCurrentUser();
            expdata.sqlexecute.Parameters.AddWithValue("@CUemployeeID", currentUser.EmployeeID);
            if (currentUser.isDelegate == true)
            {
                expdata.sqlexecute.Parameters.AddWithValue("@CUdelegateID", currentUser.Delegate.EmployeeID);
            }
            else
            {
                expdata.sqlexecute.Parameters.AddWithValue("@CUdelegateID", DBNull.Value);
            }
            expdata.ExecuteProc("restoreDefaultTooltip");
            expdata.sqlexecute.Parameters.Clear();
        }

        /// <summary>
        /// Replaces line breaks from the tooltip help text when loading for edit from the DB.
        /// </summary>
        public string ReplaceLineBreak(string inputText)
        {
            string outputText;
            outputText = inputText.Replace("<br />", System.Environment.NewLine);
            return outputText;
        }

        /// <summary>
        /// Replaces environment NewLine (\r\n) from the tooltip help text when saving back to the DB. 
        /// </summary>
        public string ReplaceNewLine(string inputText)
        {
            string outputText;
            outputText = inputText.Replace(System.Environment.NewLine, "<br />");
            return outputText;
        }
	}

	public class cHelpItem
	{		
        private string sPage;
		private string sDescription;
		private string sHelptext;
        private Guid gToolTipID;
        private string sTooltipArea;

		public cHelpItem (string page, string description, string helptext, Guid toolTipID, string tooltipArea, int? moduleID)
		{			
            sPage = page;
			sDescription = description;
			sHelptext = helptext; //.Replace("'","\\'");
            gToolTipID = toolTipID;
            sTooltipArea = tooltipArea;
            this.ModuleID = moduleID;
		}

		#region properties
        public string page
        {
            get { return sPage; }
        }
		public string description
		{
			get {return sDescription;}
		}
		public string helptext
		{
			get {return sHelptext;}
		}

        public Guid ToolTipID
        {
            get { return gToolTipID; }
        }

        public string ToolTipArea
        {
            get { return sTooltipArea; }
        }

        /// <summary>
        /// Which module this tooltip belongs, null if it is shared
        /// </summary>
        public int? ModuleID { get; set; }
		#endregion
	}
}

File: frameworkWebApp/code/cHelp.vb
Imports SpendManagementLibrary
Imports Spend_Management
Imports FWClasses
Imports System.Data
Imports System.Configuration
Imports System.Web.Caching

Public Class cHelp
    Public list As System.Collections.SortedList

    Public Cache As System.Web.Caching.Cache = CType(System.Web.HttpContext.Current.Cache, System.Web.Caching.Cache)
    Private db As New cFWDBConnection
    Private strsql As String
    Private currentUser As CurrentUser
    Private helpKey As String

    Public Sub New(ByVal curUser As CurrentUser)
        currentUser = curUser
        helpKey = currentUser.Account.companyid
        CreateDependency()
        InitialiseData()
    End Sub

    Public Sub New()
        helpKey = ""
        CreateDependency()
        InitialiseData()
    End Sub

    Private Sub CreateDependency()
        If GlobalVariables.GetAppSettingAsBoolean("EnableBrokers") AndAlso Cache("helpdependency_" & helpKey) Is Nothing Then
            Cache.Insert("helpdependency_" & helpKey, 1)
        End If
    End Sub

    Private Function getDependency() As System.Web.Caching.CacheDependency
        Dim dependency(0) As String
        dependency(0) = "helpdependency_" & helpKey

        Dim dep As New System.Web.Caching.CacheDependency(Nothing, dependency)

        Return dep
    End Function

    Public Sub InvalidateCache()
        Cache.Remove("helpdependency_" & helpKey)
        CreateDependency()
    End Sub

    Private Sub InitialiseData()
        If Cache("help_" & helpKey) Is Nothing Then
            list = CacheList()
        Else
            list = CType(Cache("help_" & helpKey), System.Collections.SortedList)
        End If
    End Sub

    Private Function CacheList() As System.Collections.SortedList
        Dim list As New System.Collections.SortedList
        Dim helpid As Guid
        Dim helptext, description, helppage As String
        Dim reqhelp As cHelpItem
        Dim db As New cFWDBConnection

        If helpKey.Trim = "" Then
            db.DBOpenMetabase(ConfigurationManager.ConnectionStrings("metabase").ConnectionString, False)
        Else
            db.DBOpen(cMigration.ConvertToFWSettings(currentUser.Account, New cAccountSubAccounts(currentUser.Account.accountid).getSubAccountsCollection, currentUser.CurrentSubAccountId), False)
        End If

        strsql = "select tooltipID, page, description, helptext from dbo.help_text"

        Using reader As System.Data.SqlClient.SqlDataReader = db.GetReader(strsql)

            While reader.Read()
                helpid = reader.GetGuid(reader.GetOrdinal("tooltipID"))
                If reader.IsDBNull(reader.GetOrdinal("helptext")) Then
                    helptext = ""
                Else
                    helptext = reader.GetString(reader.GetOrdinal("helptext"))
                End If
                If reader.IsDBNull(reader.GetOrdinal("page")) Then
                    helppage = ""
                Else
                    helppage = reader.GetString(reader.GetOrdinal("page"))
                End If
                If reader.IsDBNull(reader.GetOrdinal("description")) Then
                    description = ""
                Else
                    description = reader.GetString(reader.GetOrdinal("description"))
                End If
                reqhelp = New cHelpItem(helpid, helppage, description, helptext)
                list.Add(helpid, reqhelp)
            End While

            reader.Close()
        End Using

        db.DBClose()

        If GlobalVariables.GetAppSettingAsBoolean("EnableBrokers") AndAlso list.Count > 0 Then
            Cache.Insert("help_" & helpKey, list, getDependency(), System.Web.Caching.Cache.NoAbsoluteExpiration, TimeSpan.FromMinutes(CInt(Caching.CacheTimeSpans.Medium)), CacheItemPriority.Default, Nothing)
        End If
        Return list
    End Function

    Public Function getHelpById(ByVal helpid As Guid) As cHelpItem
        Dim tmpHI As cHelpItem = CType(list(helpid), cHelpItem)
        If tmpHI Is Nothing Then
            tmpHI = New cHelpItem(Guid.Empty, "", "", "")
        End If

        Return tmpHI
    End Function

    Public Function getGrid() As DataTable
        Dim tbl As New DataTable
        tbl.Columns.Add("helpid", GetType(System.Int32))
        tbl.Columns.Add("page", GetType(System.String))
        tbl.Columns.Add("description", GetType(System.String))
        tbl.Columns.Add("text", GetType(System.String))

        Dim includeLogonTooltips As Boolean = True
        Dim accounts As New cAccounts()

        If accounts.CachedAccounts.Count > 1 Then
            includeLogonTooltips = False
        End If

        For Each item As cHelpItem In list.Values
            If includeLogonTooltips = False And item.page = "Logon" Then
                ' skip logon tooltips for hosted systems
            Else
                tbl.Rows.Add(New Object() {item.helpid, item.page, item.description, item.helptext})
            End If
        Next
        Return tbl
    End Function

    Public Function containsTip(ByVal id As Guid) As Boolean
        Dim item As cHelpItem = getHelpById(id)
        If item.helpid = Guid.Empty Then
            Return False
        End If

        If item.helptext.Trim.Length = 0 Then
            Return False
        End If

        Return True
    End Function

    Public Sub saveTooltip(ByVal helpid As Guid, ByVal text As String)
        Dim curItem As cHelpItem = getHelpById(helpid)
        curItem.helptext = text

        Dim db As New cFWDBConnection
        db.DBOpen(cMigration.ConvertToFWSettings(currentUser.Account, New cAccountSubAccounts(currentUser.Account.accountid).getSubAccountsCollection, currentUser.CurrentSubAccountId), False)
        db.AddDBParam("helpid", helpid, True)
        db.AddDBParam("text", text, False)
        db.ExecuteSQL("EXEC saveTooltip @helpid, @text")
        db.DBClose()
    End Sub

    Public Sub restoreTooltip(ByVal helpid As Guid)
        Dim db As New cFWDBConnection
        db.DBOpen(cMigration.ConvertToFWSettings(currentUser.Account, New cAccountSubAccounts(currentUser.Account.accountid).getSubAccountsCollection, currentUser.CurrentSubAccountId), False)
        db.AddDBParam("helpid", helpid, True)
        db.ExecuteSQL("EXEC restoreDefaultTooltip @helpid")
        db.DBClose()

        InvalidateCache()
    End Sub
End Class

Public Class cHelpItem
    Private gHelpId As Guid
    Private sDescription As String
    Private sHelptext As String
    Private sPage As String

    Public Sub New(ByVal helpid As Guid, ByVal helppage As String, ByVal description As String, ByVal helptext As String)
        gHelpId = helpid
        sPage = helppage
        sDescription = description
        sHelptext = helptext.Replace("'", "`")
    End Sub

#Region "properties"
    Public ReadOnly Property helpid() As Guid
        Get
            Return gHelpId
        End Get
    End Property

    Public ReadOnly Property page() As String
        Get
            Return sPage
        End Get
    End Property

    Public ReadOnly Property description() As String
        Get
            Return sDescription
        End Get
    End Property

    Public Property helptext() As String
        Get
            Return sHelptext
        End Get
        Set(ByVal value As String)
            sHelptext = value
        End Set
    End Property
#End Region

End Class
